### DISCUSSION TITLE
May I know the meaning of this line?

### DISCUSSION THREAD

**Discussion by berttggg** [Toni-SM/skrl#126] [Q&A] [answered]
To my best understanding, `self.env.agents` is a list of names of all current agents, so `if not self.env.agents` will always return `False`?

https://github.com/Toni-SM/skrl/blob/7b090a1674b3ff1240a49385bd7fd757571f7145/skrl/trainers/torch/base.py#L311C24-L311C39

**Comment by Toni-SM** [comment, marked as answer]
Hi @berttggg 

**skrl** multi-agent environment wrapper is based on PettingZoo Parallel API.
Then, there are 2 properties related to the available agents:
* `possible_agents`: A list of the names of all possible agents the environment could generate.. This cannot be changed through play or resetting.
* `agents`: A list of the names of all current agents. These may be changed as an environment progresses (i.e. agents can be added or removed). 

The second one is used as condition to reset the environment as shown in the PettingZoo Parallel API [Usage](https://pettingzoo.farama.org/api/parallel/#usage). 

https://github.com/Farama-Foundation/PettingZoo/blob/f0c94c6f78a0d24a85d8e60f3685b076b4aa6510/docs/api/parallel.md?plain=1#L29

```python
while env.agents:
```

Think of it as a game in which, as agents (players) lose, they are removed from the game (dropped from the list). Once all players lose (the list if empty), it is necessary to restart the game again :)
